Mathlib/NumberTheory/Ostrowski.lean

```diff
@@ -55,4 +57,213 @@ lemma equiv_on_Nat_iff_equiv : (∃ c : ℝ, 0 < c ∧ (∀ n : ℕ , (f n)^c =
       (apply_nonneg f _), h x.den, ← MulRingNorm.apply_natAbs_eq,← MulRingNorm.apply_natAbs_eq,
       h (natAbs x.num)]
 
+/-!
+Throughout this file, `f` is an arbitrary absolute value.
+-/
+variable {f : MulRingNorm ℚ}
+
+open Rat.MulRingNorm
+
+section Non_archimedean
+
+-- ## Non-archimedean case
+
+/-- The mulRingNorm corresponding to the p-adic norm on ℚ. -/
+def mulRingNorm_padic (p : ℕ) [hp : Fact (Nat.Prime p)] : MulRingNorm ℚ :=
+{ toFun     := fun x : ℚ ↦ (padicNorm p x : ℝ),
+  map_zero' := by simp only [padicNorm.zero, Rat.cast_zero]
+  add_le'   := by simp only; norm_cast; exact fun r s ↦ padicNorm.triangle_ineq r s
+  neg'      := by simp only [eq_self_iff_true, forall_const, padicNorm.neg];
+  eq_zero_of_map_eq_zero' := by
+    simp only [Rat.cast_eq_zero]
+    apply padicNorm.zero_of_padicNorm_eq_zero
+  map_one' := by simp only [ne_eq, one_ne_zero, not_false_eq_true, padicNorm.eq_zpow_of_nonzero,
+    padicValRat.one, neg_zero, zpow_zero, Rat.cast_one]
+  map_mul' := by simp only [padicNorm.mul, Rat.cast_mul, forall_const]
+}
+
+@[simp] lemma mulRingNorm_eq_padic_norm (p : ℕ) [Fact (Nat.Prime p)] (r : ℚ) :
+  mulRingNorm_padic p r = padicNorm p r := rfl
+
+-- ## Step 1: define `p = smallest n s. t. 0 < |n| < 1`
+
+variable (hf_nontriv : f ≠ 1) (bdd : ∀ n : ℕ, f n ≤ 1)
+
+/-- There exists a minimal positive integer with absolute value smaller than 1 -/
+lemma exists_smallest_Nat_zero_lt_mulRingNorm_lt_one : ∃ (p : ℕ), (0 < f p ∧ f p < 1) ∧
+    ∀ (m : ℕ), 0 < f m ∧ f m < 1 → p ≤ m := by
+  -- There is a positive integer with absolute value different from one
+  have hn : ∃ (n : ℕ), n ≠ 0 ∧ f n ≠ 1 := by
+    by_contra! h
+    apply hf_nontriv
+    apply eq_on_Nat_iff_eq.1
+    intro n
+    by_cases hn0 : n=0
+    · simp only [hn0, CharP.cast_eq_zero, map_zero]
+    · push_neg at hn0
+      simp only [MulRingNorm.apply_one, Nat.cast_eq_zero, hn0, ↓reduceIte, h n hn0]
+  obtain ⟨n, hn1, hn2⟩ := hn
+  set P := {m : ℕ | 0 < f ↑m ∧ f ↑m < 1} -- p is going to be the minimum of this set
+  have hP : Set.Nonempty P := by
+    use n
+    exact ⟨map_pos_of_ne_zero f (Nat.cast_ne_zero.mpr hn1), lt_of_le_of_ne (bdd n) hn2⟩
+  use sInf P
+  exact ⟨Nat.sInf_mem hP, fun m ↦ fun hm ↦ Nat.sInf_le hm⟩
```

Comment:
```suggestion
  have hP : P.Nonempty :=
    ⟨n, map_pos_of_ne_zero f (Nat.cast_ne_zero.mpr hn1), lt_of_le_of_ne (bdd n) hn2⟩
  exact ⟨sInf P, Nat.sInf_mem hP, fun m hm ↦ Nat.sInf_le hm⟩
```
`use`, followed by `exact` can usually be combined into a single `exact` (and `by exact` cancels in most cases).